### ISSUE TITLE
Backpropagation over location network

### ISSUE THREAD

**Issue by fabienbaradel** [zhongwen/RAM#3]
Hi @zhongwen  and thank you for sharing your code. I ran your code and it gets good results.
However when I try to print the gradient for each variable I realized that there is no gradient for the variables in the location network ("loc_net").

Here is my code to print:
def print_var_and_grad(loss):
    print("\nVar and their associated grad:")
    for v in tf.global_variables():
        var_name = v.name
        var_grad = tf.gradients(loss, v)[0]
        var_grad = var_grad if var_grad is None else var_grad.name
        print("%s _______  %s" % (var_name, var_grad))

print_var_and_grad(loss)

And the log:
Var and their associated grad:
glimpse_net/Variable:0 _______  gradients_1/AddN_19:0
glimpse_net/Variable_1:0 _______  gradients_2/AddN_19:0
glimpse_net/Variable_2:0 _______  gradients_3/AddN_19:0
glimpse_net/Variable_3:0 _______  gradients_4/AddN_19:0
glimpse_net/Variable_4:0 _______  gradients_5/AddN_19:0
glimpse_net/Variable_5:0 _______  gradients_6/AddN_19:0
glimpse_net/Variable_6:0 _______  gradients_7/AddN_19:0
glimpse_net/Variable_7:0 _______  gradients_8/AddN_19:0
loc_net/Variable:0 _______  None
loc_net/Variable_1:0 _______  None
rnn_decoder/LSTMCell/W_0:0 _______  gradients_11/AddN_19:0
rnn_decoder/LSTMCell/B:0 _______  gradients_12/AddN_19:0
baseline/Variable:0 _______  gradients_13/AddN:0
baseline/Variable_1:0 _______  gradients_14/AddN:0
cls/Variable:0 _______  gradients_15/xw_plus_b_10/MatMul_grad/MatMul_1:0
cls/Variable_1:0 _______  gradients_16/xw_plus_b_10_grad/BiasAddGrad:0

Variables of the loc_net are always the same during the training to you are choice of the net location is always a random choice.

It seems I missed something.
In my point of view you just have to remove the 'stop_gradient' to make the variables of the loc_net trainable.
Could you explain me your point of view.

Regards,

**Comment by ereday** [comment]
It would be very nice to get some insight about the implementation.

**Comment by sharpstill** [comment]
I think there is a bug in the source code about the location net , just see another implementation https://github.com/jtkim-kaist/ram_modified/blob/master/ram_modified.py
He seems fix the bug?

**Comment by XoriieInpottn** [comment]
Thanks for your code! 

I think use tf.stop_gradient() for both "mean_loc" and "sample_loc" causes the gradients of location network to be None.
So tf.stop_gradient(mean_loc) should be in "get_next_input()". Please check.

Many implementations in github such as:
https://github.com/zhongwen/RAM
https://github.com/jlindsey15/RAM
https://github.com/hehefan/Recurrent-Attention-Model
have the same problem.
